base/parse-type.jl

```diff
@@ -0,0 +1,208 @@
+            @_bail_if_nothing type_var = _try_parse_type_var(module_context, ast.args[i], raise, type_vars)
+            type_var::TypeVar
+            type_vars[type_var.name] = type_var
+            push!(new_type_vars, type_var)
+        end
+        # Then evaluate the body in the context of those type vars
+        @_bail_if_nothing body = _try_parse_type(module_context, ast.args[1], raise, type_vars)
+        # Now work backwards through the new type vars and construct our wrapper UnionAlls:
+        for type_var in Iterators.reverse(new_type_vars)
+            body = UnionAll(type_var, body)
+        end
+        return body
+    elseif ast isa Expr && ast.head == :call && ast.args[1] === :typeof
+        return typeof(@_bail_if_nothing _try_parse_type(module_context, ast.args[2], raise, type_vars))
+    elseif ast isa Expr && ast.head == :call
+        return @_bail_if_nothing _try_parse_isbits_constructor(module_context, ast, raise, type_vars)
+    else
+        return @_bail_if_nothing _try_parse_qualified_type(module_context, ast, raise, type_vars)
+    end
+end
+_try_parse_qualified_type(module_context, val, _, _) = val
+function _try_parse_qualified_type(module_context, ast::Expr, raise::Bool, type_vars)
+    if ast.head !== :(.)
+        raise && throw(ArgumentError("Failed to parse type expression. Expected a \
+                qualified type, e.g. `Base.Dict`, got: `$ast`"))
+        return nothing
+    end
+    @_bail_if_nothing mod = _try_parse_qualified_type(module_context, ast.args[1], raise, type_vars)
+    value = ast.args[2]
+    if value isa QuoteNode
+        value = value.value
+    end
+    return getglobal(mod, value)
+end
+function _try_parse_qualified_type(module_context, sym::Symbol, raise::Bool, type_vars)
+    # First try to look up the symbol in the type vars
+    if type_vars !== nothing
+        v_if_found = get(type_vars, sym, :not_found)
+        if v_if_found !== :not_found
+            return v_if_found
+        end
+    end
+
+    # Otherwise, this should be a top-level name in the provided module.
+
+    if !raise  # Skip the isdefined check if !raise, since getglobal will throw anyway.
+        !isdefined(module_context, sym) && return nothing
+    end
+    getglobal(module_context, sym)
+end
+
+# Parses constant isbits constructor expressions, like `Int32(10)` or `Point(0,0)`, as used in type
+# parameters like `Val{10}()` or `DefaultDict{Point(0,0)}`.
+function _try_parse_isbits_constructor(module_context, ast, raise::Bool, type_vars)
+    @_bail_if_nothing typ = _try_parse_type(module_context, ast.args[1], raise, type_vars)
+    # PERF: Reuse the args vector when parsing the type values.
+    popfirst!(ast.args)
+    for i in 1:length(ast.args)
+        @_bail_if_nothing ast.args[i] = _try_parse_type(module_context, ast.args[i], raise, type_vars)
+    end
+    # We use reinterpret to avoid evaluating code, which may have side effects.
+    return reinterpret(typ, Tuple(ast.args))
```

Comment:
This assumes the default constructor. It can be wrong, though

```julia
julia> struct Point
           x::Int32
           y::Int32
           Point(x, y=x) = new(x, y)
       end

julia> str = "Val{Point(-7)}"
"Val{Point(-7)}"

julia> eval(Meta.parse(str))
Val{Point(-7, -7)}

julia> parse(Type, str)
Val{Point(-7, -1)}
```

The task of constructing arbitrary values at runtime without evaluating code is tricky.

Reply:
Yeah, interesting... 🤔 ..... so maybe this whole idea is not valid? :(
Maybe we could get away with it by dropping values, and documenting a restriction that we don't support struct values in the type domain, and for those, we recommend falling back to `eval`? 

:( Sadbeans! Thanks for catching this. Do you have any other good ideas / suggestions?

Reply:
If we can't get this right, maybe this ultimately doesn't belong in base, and it should be a library similar to what @tecosaur shared.

Reply:
But also also, i was mainly interested in using this for improving PackageCompiler, and it's worth noting that structs in types are currently broken in precompile statements anyway, since they're printed as kwargs for some reason!?:
```
precompile(Tuple{Type{Base.Val{Main.Point(x=-7, y=-7)}}})
```
```julia
julia> str = "precompile(Tuple{Type{Base.Val{Main.Point(x=-7, y=-7)}}})"
"precompile(Tuple{Type{Base.Val{Main.Point(x=-7, y=-7)}}})"

julia> eval(Meta.parse(str))
ERROR: MethodError: no method matching Point(; x::Int64, y::Int64)
```
So these cases are all broken anyway, so maybe we don't need to support them? (though i'd like to fix the way they're printed in precompiles, too...)

Reply:
For a `parse` method, I think it's reasonable to only support values in type parameters if the values are embedded directly in the syntax tree (e.g. `Val{5}` or `Val{"a"}` but not `Val{1+1}` or `Val{(1,1)}`.

Reply:
IDK if that will be enough for your use-case, though.

Reply:
I definitely agree I wouldn't support `Val{1+1}`, since that is definitely _not_ a value embedded directly in the tree.

But isn't `Val{(1,1)}`? I do think we see constant tuples in our types fairly frequently, so i would want that.

Reply:
Maybe special case to allow this iff the default inner constructor is accessible and the parsed string would dispatch to it?